Download plugins from main plugins option page

zas opened the pull request:
Still need to be polished, but it removed the download plugins sub-page, and simplify the code.

Also permit to install zipped plugins from local filesystems.

dufferzafar committed:
Adding download plugin functionality to existing UI

A list of plugins (JSON) is downloaded when Picard starts. The JSON
data is used to match already present plugins with those on the server.

The result is something like this: http://i.imgur.com/cL8VuM5.jpg

The entries checked are the plugins that are enabled. The ones in the
bold are those which are available for download, i.e, they are
present on our Github Repo. Those not bold might have been added by the
user from other means (such as by copying plugin files to Picard's
directory.)

One thing that worries me about the bold ones is that as plugins get
added to the GH repo, most of the entries in the list will be bold, and
that won't look good.

All the user has to do is select one in the bold and press the "Download
Plugin" button (that part isn't finished yet.)

The data used to check if the plugins are on our Github repo or not is here:
https://raw.githubusercontent.com/dufferzafar/picard-plugins/master/Plugins.json

The script used to generate the JSON and a API that helps with downloads are here:
https://github.com/dufferzafar/picard-plugins

Apart from the actual download, the one thing that still needs to be
done is listing the plugins that are on Github but not present in plugin
folders.

dufferzafar committed:
Added a new options page to download plugins

This is a WIP.

Trivial things (such as chage_details) in the new download
UI have been implemented.

Todo: Use the json provided by /api rather than Github.
Todo: Select the first item in the download plugins list.

dufferzafar committed:
Things don't break if the json file is not available on disk

The Download treeview just remains empty (as it should?)

dufferzafar committed:
Plugins with an update available are shown in bold

The version available can also be added to the details.

The download plugin button will renamed to update plugin and will be
enabled only when needed.

dufferzafar committed:
Code now uses the API and the new json structure

JSON data is downloaded from our API rather than a static file on
Github.

The code is now based on the new json structure which stores the plugins
as a dictionary (with module_name as key.)

dufferzafar committed:
Conform to the updated API

...which has a versioned URL and outputs plugins correctly.

dufferzafar committed:
Added download and install capability

The plugin is downloaded from the /download endpoint of the API which
increases the download count and serves the plugin zip files.

The zip is downloaded and extracted to USER_PLUGIN_DOWNLOAD_DIR.

The extracted plugin is then installed (loaded, but not enabled.)
Loading the plugin moves it to USER_PLUGIN_DIR.

The zip is then deleted but the extracted folder contents are kept,
I don't have any specific reason to keep the extracted folder so, this
can be deleted too...

TODO:

    If the plugin couldn't be downloaded (network error or w/e) a dialog
    box is shown, but If it was downloaded and installed, the user does
    not get any notification.

    Add a checkbox "Hide alread installed plugins." - which can be used
    to view plugins that are available but not installed.

dufferzafar committed:
Refactored loading of plugins list

Previously, the data was first written to a file and then read again
by different options pages - which was just stupid.

Now, the data is loaded as json right when its download.
All other pages can then just use it.

A statusbar message is set if an error occurs in loading the list.

dufferzafar committed:
Show version installed in the details

So it serves as a notifier that the plugin has been successfully
installed.

We could probably have a message box or a status bar message too.

dufferzafar committed:
The install plugin button is selectively enabled

...for plugins that are not already installed.

For the plugins that are installed, maybe we could turn the button to
'Update plugin', so it updates the plugin (if an update is available)

dufferzafar committed:
Added a 'Hide installed plugins' checkbox to the downloads page

We could have a setting that defines the default state of the checkbox.

The checkbox feels a bit oddly placed, a push button would've been
better?

dufferzafar committed:
The Update plugin button is selectively enabled

...only for plugins which have an update available.

The 'Download plugin' button which used to open the doc page
has been replaced by this update plugin button.

Todo:

    1. Check if the local plugin version has any changes.

    2. Download and Install.

dufferzafar committed:
Updated constants

Changed USER_PLUGIN_DOWNLOAD_DIR to USER_DOWNLOADS_DIR

Moved the API URLs to PICARD_URLS dictionary

Added a URL for the plugins repository

dufferzafar committed:
Addressed some issues from comments

Renamed the variable 'pluginjson' to 'plugins_available'

Removed checkboxes from the plugins download list.

Renamed the file 'plugins_dl.py' to 'plugins_download.py'

dufferzafar committed:
Installed plugins are not shown in the available list

This keeps things clear - two lists for distinct purposes - Installed &
Available.

The "Hide installed plugins" checkbox has been removed and the groupboxes
renamed.

dufferzafar committed:
Use the webservice to download data asynchronously

The plugins list and all the plugins are downloaded via this method.

A msgbox is showed once a plugin is successfully installed, and is
removed from the available list.

dufferzafar committed:
Added Update plugin capability

Before updating, the user is warned if the local plugin files have been
edited.

The only thing that's left now is to move the common methods b/w
plugins.py and plugins_download.py to a single file.

zas committed:
Update code, introduce PLUGINS_API dict to store host/port/endpoints

zas committed:
Add license, license_url properties to PluginWrapper

zas committed:
Display plugins license

zas committed:
Add a 'new_version' property to PluginWrapper

Use it instead of testing for 'bold'.

zas committed:
Mark dialog strings for translation

zas committed:
Remove 'Downloads' field, not anymore provided by plugins JSON API

zas committed:
Support plugin install from local zip archives

zas committed:
Remove now unused import

zas committed:
Move available plugins query code to plugin manager

zas committed:
Fix downloaded plugin install

zas committed:
Add an Update button for each updateable plugin

Simplify code since we can now install directly from zip files.

zas committed:
Enable download from main plugin page, drop sub page

zas committed:
Remove Update plugin button

dufferzafar commented:
On first run, I got this:

```
Traceback (most recent call last):
  File "./picard/ui/options/plugins.py", line 171, in change_details
    text.append("<b>" + _("Files") + "</b>: " + plugin.files_list)
AttributeError: 'PluginWrapper' object has no attribute 'files_list'
Traceback (most recent call last):
  File "./picard/ui/options/plugins.py", line 171, in change_details
    text.append("<b>" + _("Files") + "</b>: " + plugin.files_list)
AttributeError: 'PluginWrapper' object has no attribute 'files_list'
Traceback (most recent call last):
  File "./picard/ui/options/plugins.py", line 171, in change_details
    text.append("<b>" + _("Files") + "</b>: " + plugin.files_list)
AttributeError: 'PluginWrapper' object has no attribute 'files_list'
```

But I can't seem to reproduce it. :/

dufferzafar commented:
I think the columns should resize to take up all the space. This is how it looks [right now](http://i.imgur.com/ybpWEcb.png).

And the 'Installed Plugins' label should be changed to 'Plugins'?

zas commented:
> On first run, I got this:
> 
> Traceback (most recent call last):
>   File "./picard/ui/options/plugins.py", line 171, in change_details
>     text.append("<b>" + _("Files") + "</b>: " + plugin.files_list)
> AttributeError: 'PluginWrapper' object has no attribute 'files_list'
> Traceback (most recent call last):
>   File "./picard/ui/options/plugins.py", line 171, in change_details
>     text.append("<b>" + _("Files") + "</b>: " + plugin.files_list)
> AttributeError: 'PluginWrapper' object has no attribute 'files_list'
> Traceback (most recent call last):
>   File "./picard/ui/options/plugins.py", line 171, in change_details
>     text.append("<b>" + _("Files") + "</b>: " + plugin.files_list)
> AttributeError: 'PluginWrapper' object has no attribute 'files_list'
> 
> But I can't seem to reproduce it. :/

Yes, i'm aware of such issue, it is related to data about plugins, on one side we have an object PluginWrapper for installed plugins and on the other side a dict from JSON API data, we need to somehow make something more common and practical, current code is somehow working, but too hacky.

zas commented:
> I think the columns should resize to take up all the space. This is how it looks right now.

Yes, but i reduced last column, because it was the one expanded, leading to BIG buttons, i'll rework that soon.

> And the 'Installed Plugins' label should be changed to 'Plugins'?

I agree.

zas committed:
Installed Plugins -> Plugins

zas committed:
Store plugin info from JSON API in PluginData objects

It helps to have a common interface for installed plugins and available plugins.

zas committed:
PLugins list: expand name column to the max size

zas committed:
Reduce height of download/update buttons to a minimum

zas committed:
Update/download buttons: simplify code

zas commented:
Please test again, i addressed most obvious issues.

dufferzafar commented:
Things seem to be working fine now.

zas committed:
Use single quotes

zas committed:
Display new version and refresh infos after plugin installation

zas committed:
Improve display of plugin details

zas committed:
Introduce class PluginShared

It is used for shared properties between PluginWrapper and PluginData

zas committed:
Introduce plugin flags and reduce number of add_plugin_item()'s args

zas commented:
PR can be merged if you are ok with it.

zas commented:
@dufferzafar : can you update your https://github.com/dufferzafar/picard/tree/test_download_ui since it is the branch tracked by https://github.com/musicbrainz/picard/pull/334 ?
